transmart-core-db/grails-app/services/org/transmartproject/db/clinical/MultidimensionalDataResourceService.groovy

```diff
@@ -787,112 +746,44 @@ class MultidimensionalDataResourceService implements MultiDimensionalDataResourc
-        }
-
-        at.collect {aggregateFieldType(it)}.unique().findAll().each {
-            def wrongValueTypeConstraint = new FieldConstraint(
-                    operator: Operator.NOT_EQUALS,
-                    field: valueTypeField,
-                    value: it,
-            )
-            if(exists(builder, new AndConstraint(args: [(Constraint) constraint, wrongValueTypeConstraint]))) {
-                throw new InvalidQueryException('One of the concepts/observations has the wrong type for this aggregation')
-            }
-        }
-
-        // check if the concept is truly numerical (all textValue are E and all numberValue have a value) or textual
-
-        def textTypeConstraint = new FieldConstraint(
-                operator: Operator.EQUALS,
-                field: valueTypeField,
-                value: ObservationFact.TYPE_TEXT,
-        )
-
-        def numberTypeConstraint = new FieldConstraint(
-                operator: Operator.EQUALS,
-                field: valueTypeField,
-                value: ObservationFact.TYPE_NUMBER,
-        )
-
-        def textValueEConstraint = new FieldConstraint(
-                operator: Operator.EQUALS,
-                field: textValueField,
-                value: "E"
-        )
-
-        // Transmart currently does not allow null values in the ObservationFacts, but it could be extended to allow
-        // that.
-        def numberValueNotNullConstraint = new Negation(arg: new NullConstraint(field: numberValueField))
-        def textValueNotNullConstraint = new Negation(arg: new NullConstraint(field: textValueField))
-
-        def invalidObservationConstraint = new Negation(arg: new OrConstraint(args: [
-                new AndConstraint(args: [numberTypeConstraint, textValueEConstraint, numberValueNotNullConstraint]),
-                new AndConstraint(args: [textTypeConstraint, textValueNotNullConstraint])
-        ]))
-
-        def invalidObservations = getIterable(builder.buildCriteria(
-                new AndConstraint(args: [(Constraint) constraint, invalidObservationConstraint])))
+    @Override
+    Map<String, Long> countCategoricalValues(MultiDimConstraint constraint, User user) {
+        assert constraint instanceof Constraint
+        checkAccess(constraint, user)
 
-        invalidObservations.withCloseable {
-            for(ObservationFact o in invalidObservations) {
-                // Retrieving the value will throw an exception
-                o.value
-            }
-        }
+        def builder = getCheckedQueryBuilder(user)
+        DetachedCriteria criteria = builder.buildCriteria(constraint)
+        def projections = Projections.projectionList()
+        projections.add(Projections.groupProperty('textValue'))
+        projections.add(Projections.rowCount())
+        criteria.setProjection(projections)
+                .setResultTransformer(Transformers.TO_LIST)
+                .add(Restrictions.eq('valueType', ObservationFact.TYPE_TEXT))
```

Comment:
Also filter out null values?

Reply:
It might be useful to know a number of missed values. We decided to keep them. `{null: 343}`. I will add test case for that.